### DISCUSSION TITLE
Is there a way to allow only the broker to publish messages?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [moscajs/aedes#678] [Q&A] [answered]
@robertsLando, a bit OT: where do you prefere to ask questions about `aedes`? In GitHub issues or GitHub discussions or StackOverflow? Could this be noted in the `README.md` please?

---

In my use case, only the broker publishes messages. Is there a way to prohibit all clients from publishing any messages? Do I need to use authentication?

**Comment by robertsLando** [comment, marked as answer]
Questions should be asked in discussions. Issues/Features in issues tracker.

About your question I suggest to read docs: https://github.com/moscajs/aedes/blob/main/docs/Aedes.md#handler-authorizepublish-client-packet-callback

**Reply by tukusejssirs** [reply]
> Questions should be asked in discussions. Issues/Features in issues tracker.

Then I think you should remove the issue template for questions and probably add a link to the discussions there with a note that the questions should be asked here. :wink:

> About your question I suggest to read docs: https://github.com/moscajs/aedes/blob/main/docs/Aedes.md#handler-authorizepublish-client-packet-callback

Okay, then how can I check if the publisher is the broker? Is the `authorizePublish()` function run only when a _client_ publishes a message or also when the broker publishes it using `aedes.publish()`?

**Reply by robertsLando** [reply]
When the publisher is the brokeer cliient is null

**Reply by tukusejssirs** [reply]
Both `return callback(new Error('wrong topic'))` and `return callback('wrong topic')` close the client connection. However, all I want to disregard the messages published by any client. I may return an error message to the client, but the connection should stay open.

How could I achieve this?

**Edit:** My bad, it was a different, unrelated error. However, how can I make sure the client.

I tried to run the following in to `node -i` sessions:

<details>
  <summary>1. Initialise the server in the first Node session.</summary>

```js
const aedes = require('aedes')()
const {createServer} = require('aedes-server-factory')
const server = createServer(aedes, {ws: true})
server.listen(8080, () => {
  console.info('[INFO] Server started and listening on port ' + 8080 + '.')
})
aedes.on('publish', (packet, client, callback) => {
  if (client === null && ! packet.topic.startsWith('$SYS/')) {
    console.log('publisher 1 : root : ' + packet.topic)
  } else if (! packet.topic.startsWith('$SYS/')) {
    console.log('non-broker publisher')
    console.log('non-broker topic : ' + packet.topic)
    console.log('client :', client)
  }
})

aedes.authorizePublish = ('publish', (packet, client) => {
  if (packet.topic.startsWith('$SYS/')) {
    callback(null)
  } else if (client === null) {
    console.log('publisher 2 : root : ' + packet.topic)
    callback(null)
  } else {
    console.log('non-broker publisher')
    console.log('non-broker topic : ' + packet.topic)
    // return callback(new Error('non-broker publisher : ' + client.id))
    // return callback('non-broker publisher : ' + client.id)

    aedes.publish({
      cmd: 'publish',
      qos: 0,
      topic: packet.topic,
      payload: Buffer.from('{"error": "MQTT clients are not allowed to publish messages."}'),
      retain: false
    }, (err) => {
      if (err) {
        console.log('Error publishing ' + packet.topic + ' (error message) :', err)
        return
      }
    })

    return callback(new Error('non-broker publisher'))
  }
})
```
</details>

<details>
  <summary>2. Initialise the client in the second Node session.</summary>

```js
const mqtt = require('mqtt')
const client = mqtt.connect('ws://localhost:8080')

client.subscribe('test/client', (err, granted) => {
  if (err) {
    console.error('err while client subscribing :', err)
    return
  }
})

client.on('message', (topic, message) => {
  console.log(topic, ':', JSON.parse(message.toString()))
})
```
</details>

<details>
  <summary>3. Publish a message from the server.</summary>

Output is `publisher 1 : root : test/broker` as expected.

```js
aedes.publish({
  cmd: 'publish',
  qos: 0,
  topic: 'test/broker',
  payload: Buffer.from('broker'),
  retain: false
}, (err) => {
  if (err) {
    console.log('Error publishing test/broker :', err)
    return
  }
})
```
</details>

<details>
  <summary>4. Publish a message from the client.</summary>

Expected output on the server side: _at least_ `non-broker topic : test/client`, but it does not show up.

Expected message on the client side: `test/client : {"error": "MQTT clients are not allowed to publish messages."}`

Actual output is none on both sides.

```js
client.publish({
  cmd: 'publish',
  qos: 0,
  topic: 'test/client',
  payload: Buffer.from('client'),
  retain: true
}, (err) => {
  if (err) {
    console.log('Error publishing test/client :', err)
    return
  }
})
```
</details>

I’d like to make sure no client can publish a message, only subscribing a predefined set of topics.
